File: src/main/java/io/myosotisdev/utopianism/serialize/Serializer.java
package io.myosotisdev.utopianism.serialize;

import net.minestom.server.tag.TagSerializer;
import org.jglrxavpok.hephaistos.nbt.NBT;
import org.jglrxavpok.hephaistos.nbt.NBTCompound;
import org.jglrxavpok.hephaistos.nbt.NBTString;

import java.util.HashMap;
import java.util.function.Function;

public abstract class Serializer<T>
{
    public TagSerializer<T> createTagSerializer(String key)
    {
        return TagSerializer.fromCompound(
                compound -> this.deserialize(compound.getString(key)),
                particle -> new NBTCompound().withEntries(NBT.Entry(key, new NBTString(this.serialize(particle))))
        );
    }

    public abstract String serialize(T obj);

    public abstract T deserialize(String string);
}

File: src/main/java/io/myosotisdev/utopianism/serialize/Serializer.kt
package io.myosotisdev.utopianism.serialize

import net.minestom.server.tag.TagSerializer
import org.jglrxavpok.hephaistos.nbt.NBT
import org.jglrxavpok.hephaistos.nbt.NBTCompound
import org.jglrxavpok.hephaistos.nbt.NBTString

abstract class Serializer<T>
{
    fun createTagSerializer(key: String?): TagSerializer<T>
    {
        return TagSerializer.fromCompound(
                { compound: NBTCompound -> deserialize(compound.getString(key!!)) }
        ) { particle: T -> NBTCompound().withEntries(NBT.Entry(key!!, NBTString(serialize(particle)!!))) }
    }

    abstract fun serialize(obj: T?): String?
    abstract fun deserialize(string: String?): T?
}